File: pure_pursuit_yb/src/pure_pursuit.cpp
// src/pure_pursuit.cpp
// ──────────────────────────────────────────────────────────────
// • 입력 1: /local_planned_path (nav_msgs/Path)
// • 입력 2: TF (reference -> base_link, base_link -> os_sensor)
//
// • 출력 1: /steering_angle (std_msgs/Float32, [deg])
// • 출력 2: /lookahead_point_marker (visualization_msgs/Marker)
// ──────────────────────────────────────────────────────────────

#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/path.hpp>
#include <std_msgs/msg/float32.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <visualization_msgs/msg/marker.hpp>

#include <vector>
#include <cmath>
#include <string>
#include <memory>
#include <limits>

using namespace std::chrono_literals;

class PurePursuit : public rclcpp::Node
{
public:
  PurePursuit()
  : Node("pure_pursuit")
  {
    // ===== 파라미터 =====
    this->declare_parameter<double>("wheelbase", 2.6);
    this->declare_parameter<double>("lookahead_distance", 3.0);
    this->declare_parameter<int>("control_frequency", 30);

    wheelbase_ = this->get_parameter("wheelbase").as_double();
    lookahead_distance_ = this->get_parameter("lookahead_distance").as_double();
    int control_freq = this->get_parameter("control_frequency").as_int();

    // TF 리스너 및 버퍼 초기화
    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    // ===== 구독자 =====
    path_sub_ = this->create_subscription<nav_msgs::msg::Path>(
      "/local_planned_path", 10,
      std::bind(&PurePursuit::pathCallback, this, std::placeholders::_1)
    );

    // ===== 퍼블리셔 =====
    steering_pub_ = this->create_publisher<std_msgs::msg::Float32>("/steering_angle", 10);
    lookahead_marker_pub_ = this->create_publisher<visualization_msgs::msg::Marker>("/lookahead_point_marker", 10);

    // ===== 제어 타이머 =====
    auto control_period = std::chrono::duration<double>(1.0 / control_freq);
    timer_ = this->create_wall_timer(
      control_period,
      std::bind(&PurePursuit::controlLoop, this)
    );

    RCLCPP_INFO(this->get_logger(), "Pure Pursuit node started.");
    RCLCPP_INFO(this->get_logger(), " - Wheelbase: %.2f m", wheelbase_);
    RCLCPP_INFO(this->get_logger(), " - Lookahead Distance: %.2f m", lookahead_distance_);
    RCLCPP_INFO(this->get_logger(), " - Control Frequency: %d Hz", control_freq);
  }

private:
  void pathCallback(const nav_msgs::msg::Path::SharedPtr msg)
  {
    current_path_ = msg;
  }

  void publishLookaheadMarker(int target_idx)
  {
    visualization_msgs::msg::Marker clear_marker;
    if (current_path_)
      clear_marker.header.frame_id = current_path_->header.frame_id;
    else
      clear_marker.header.frame_id = "os_sensor"; // 기본 프레임

    clear_marker.header.stamp = this->now();
    clear_marker.ns = "pure_pursuit_lookahead";
    clear_marker.action = visualization_msgs::msg::Marker::DELETEALL;
    lookahead_marker_pub_->publish(clear_marker);

    if (target_idx < 0) return;

    visualization_msgs::msg::Marker lookahead_marker;
    lookahead_marker.header.frame_id = current_path_->header.frame_id;
    lookahead_marker.header.stamp = this->now();
    lookahead_marker.ns = "pure_pursuit_lookahead";
    lookahead_marker.id = 0;
    lookahead_marker.type = visualization_msgs::msg::Marker::SPHERE;
    lookahead_marker.action = visualization_msgs::msg::Marker::ADD;

    lookahead_marker.pose.position = current_path_->poses[target_idx].pose.position;
    lookahead_marker.pose.orientation.w = 1.0;

    lookahead_marker.scale.x = 0.5;
    lookahead_marker.scale.y = 0.5;
    lookahead_marker.scale.z = 0.5;

    // 색상: 마젠타
    lookahead_marker.color.r = 1.0f;
    lookahead_marker.color.g = 0.0f;
    lookahead_marker.color.b = 1.0f;
    lookahead_marker.color.a = 1.0f;

    lookahead_marker.lifetime = rclcpp::Duration::from_seconds(0.2);
    lookahead_marker_pub_->publish(lookahead_marker);
  }

  void controlLoop()
  {
    if (!current_path_ || current_path_->poses.empty()) {
      publishLookaheadMarker(-1);
      RCLCPP_WARN_ONCE(this->get_logger(), "Waiting for path...");
      return;
    }

    geometry_msgs::msg::TransformStamped t;
    const std::string& path_frame = current_path_->header.frame_id;
    try {
      t = tf_buffer_->lookupTransform("base_link", path_frame, tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_ERROR(this->get_logger(), "Could not transform %s to base_link: %s", path_frame.c_str(), ex.what());
      return;
    }

    // 경로를 base_link로 변환
    std::vector<geometry_msgs::msg::Point> transformed_path;
    transformed_path.reserve(current_path_->poses.size());
    for (const auto& pose_stamped : current_path_->poses) {
      geometry_msgs::msg::PoseStamped transformed_pose;
      tf2::doTransform(pose_stamped, transformed_pose, t);
      transformed_path.push_back(transformed_pose.pose.position);
    }

    // 룩어헤드 타겟 인덱스 선택
    int target_idx = -1;
    for (size_t i = 0; i < transformed_path.size(); ++i) {
      if (transformed_path[i].x > 0) {
        double dist = std::hypot(transformed_path[i].x, transformed_path[i].y);
        if (dist >= lookahead_distance_) {
          target_idx = static_cast<int>(i);
          break;
        }
      }
    }
    if (target_idx == -1 && !transformed_path.empty()) {
      double max_dist = -1.0;
      for (size_t i = 0; i < transformed_path.size(); ++i) {
        if (transformed_path[i].x > 0) {
          double d = std::hypot(transformed_path[i].x, transformed_path[i].y);
          if (d > max_dist) {
            max_dist = d;
            target_idx = static_cast<int>(i);
          }
        }
      }
    }

    publishLookaheadMarker(target_idx);

    if (target_idx == -1) {
      RCLCPP_WARN(this->get_logger(), "No valid target point found in front of the vehicle.");
      return;
    }

    // Pure Pursuit 조향각 계산
    double target_x = transformed_path[target_idx].x;
    double target_y = transformed_path[target_idx].y;

    double alpha = std::atan2(target_y, target_x);
    double steering_rad = std::atan2(2.0 * wheelbase_ * std::sin(alpha), lookahead_distance_);
    double steering_deg = steering_rad * 180.0 / M_PI;

    std_msgs::msg::Float32 steering_msg;
    steering_msg.data = static_cast<float>(steering_deg);
    steering_pub_->publish(steering_msg);
  }

  // 멤버 변수
  double wheelbase_;
  double lookahead_distance_;

  rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr path_sub_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr steering_pub_;
  rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr lookahead_marker_pub_;
  rclcpp::TimerBase::SharedPtr timer_;

  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

  nav_msgs::msg::Path::ConstSharedPtr current_path_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PurePursuit>());
  rclcpp::shutdown();
  return 0;
}

File: purepursuit_k/purepursuit_k/purepursuit.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32
import math
import numpy as np
from nav_msgs.msg import Path
from visualization_msgs.msg import Marker
from std_msgs.msg import ColorRGBA
from geometry_msgs.msg import Point
from rclpy.time import Time
from tf2_ros import Buffer, TransformListener
from geometry_msgs.msg import TransformStamped

#---- Vehicle Params ----
MAX_STEERING = 30      # [deg]
CAR_POSITION = (0.0, 0.0)
PATH_INTERVAL = 0.2
SPEED_2_RPM = 0.00535  # m/s -> rpm
LAD_MIN = 4.2
LAD_MAX = 10.0
LAD_MINMAX_THRESH = 7.0


class PurePursuitControl(Node):
    def __init__(self):
        super().__init__('PurePursuitControl')

        # Params
        self.declare_parameter('WHEELBASE', 1.295)
        self.declare_parameter('SPEED', 20.0)
        self.declare_parameter('target_frame', 'base_link')
        self.declare_parameter('LAD_MIN', LAD_MIN) #초기값
        self.declare_parameter('LAD_MAX', LAD_MAX) #초기값

        self.WHEELBASE = float(self.get_parameter('WHEELBASE').value)
        self.SPEED = float(self.get_parameter('SPEED').value)
        self.target_frame = str(self.get_parameter('target_frame').value)
        self.LAD_MIN = float(self.get_parameter('LAD_MIN').value)
        self.LAD_MAX = float(self.get_parameter('LAD_MAX').value)

        # Subscribers
        self.create_subscription(Path, '/local_planned_path', self.callback_path, 10)
        self.create_subscription(Float32, '/current_speed', self.callback_currentspeed, 10)

        # Publishers
        self.pub_rpm = self.create_publisher(Float32, '/cmd/rpm', 10)
        self.pub_steer = self.create_publisher(Float32, '/cmd/steer', 10)

        # Visualize
        self.pub_lad = self.create_publisher(Marker, '/lookahead_point_marker', 1)
        self.pub_target = self.create_publisher(Marker, '/pp_target_marker', 1)
        self.pub_line = self.create_publisher(Marker, '/pp_target_line', 1)

        # State
        self.path = []
        self.current_speed = 0.0

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        # Timer
        self.timer = self.create_timer(0.1, self.main_loop)

        self.get_logger().info('----- Pure Pursuit Control Started (Variable LAD) -----')

    def callback_path(self, msg: Path):
        path_xy = []
        poses = msg.poses
        if not poses:
            self.path = []
            return

        src_frame = msg.header.frame_id if msg.header.frame_id else self.target_frame

        last = None
        for ps in poses:
            x = float(ps.pose.position.x)
            y = float(ps.pose.position.y)
            if not (math.isfinite(x) and math.isfinite(y)):
                continue

            if src_frame != self.target_frame:
                out = self._transform_xy(x, y, src_frame, self.target_frame)
                if out is None:
                    continue
                x, y = out

            if last is None or math.hypot(x - last[0], y - last[1]) >= PATH_INTERVAL:
                path_xy.append((x, y))
                last = (x, y)

        if src_frame != self.target_frame and path_xy:
            self.get_logger().info(
                f"Transformed path from {src_frame} to {self.target_frame} (N={len(path_xy)})"
            )

        self.path = path_xy

    def callback_currentspeed(self, msg: Float32):
        self.current_speed = float(msg.data)

    def _quat_to_yaw(self, q):
        siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
        cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
        return math.atan2(siny_cosp, cosy_cosp)

    def _transform_xy(self, x, y, src_frame, target_frame):
        try:
            tf: TransformStamped = self.tf_buffer.lookup_transform(
                target_frame, src_frame, Time())
        except Exception as e:
            self.get_logger().warn(f"TF lookup failed {src_frame}->{target_frame}: {e}")
            return None

        t = tf.transform.translation
        q = tf.transform.rotation
        yaw = self._quat_to_yaw(q)
        c, s = math.cos(yaw), math.sin(yaw)

        xb = c * x - s * y + t.x
        yb = s * x + c * y + t.y
        return (float(xb), float(yb))

    # ★★★ 변경된 함수: 가변 LAD 계산 로직 ★★★
    def calculate_variable_lad(self):
        if len(self.path) < 2:
            return self.LAD_MIN

        angles = []
        for i in range(1, len(self.path)):
            p_prev = self.path[i-1]
            p_curr = self.path[i]
            dx = p_curr[0] - p_prev[0]
            dy = p_curr[1] - p_prev[1]
            angle = abs(math.atan2(dy, dx))
            angles.append(abs(angle))

        if not angles:
            return self.LAD_MIN

        avg_angle = sum(angles) / len(angles)

        normalized_angle = avg_angle / (math.pi / 4.0)
        lad = self.LAD_MAX - (normalized_angle * (self.LAD_MAX - self.LAD_MIN))
        lad = float(np.clip(lad, self.LAD_MIN, self.LAD_MAX))
        lad = LAD_MAX if lad > LAD_MINMAX_THRESH else LAD_MIN
        return lad

    def find_target(self, lad: float):
        if not self.path:
            self.get_logger().warn("Path is empty. Returning (0.0, 0.0) as target.")
            return (0.0, 0.0)
        target = None
        min_err = float('inf')
        for (x, y) in self.path:
            distance = math.hypot(x, y)
            err = abs(lad - distance)
            if err < min_err:
                min_err = err
                target = (x, y)
        return target if target is not None else (0.0, 0.0)

    def compute_steering(self, target):
        dx = target[0] - CAR_POSITION[0]
        dy = target[1] - CAR_POSITION[1]
        alpha = math.atan2(dy, dx)
        steering = -1 * math.atan2(2.0 * self.WHEELBASE * math.sin(alpha), math.hypot(dx, dy))
        return steering

    def compute_speed2rpm(self, speed):
        rpm = (speed / 3.6) / SPEED_2_RPM  # km/h -> m/s -> RPM
        return rpm

    def drive(self, steering, speed):
        steer_msg = Float32()
        steer_msg.data = float(steering)
        self.pub_steer.publish(steer_msg)

        rpm_msg = Float32()
        rpm_msg.data = float(speed)
        self.pub_rpm.publish(rpm_msg)

    def vis_lad(self, lad: float):
        if lad is None or lad <= 0:
            return
        marker = Marker()
        marker.header.frame_id = 'base_link'
        marker.header.stamp = self.get_clock().now().to_msg()
        marker.ns = 'lad_semicircle'
        marker.id = 1
        marker.type = Marker.LINE_STRIP
        marker.action = Marker.ADD
        marker.scale.x = 0.03
        marker.color = ColorRGBA(r=0.0, g=1.0, b=1.0, a=1.0)
        marker.pose.orientation.w = 1.0
        cx, cy = CAR_POSITION
        num_pts = 64
        for theta in np.linspace(-math.pi/2, math.pi/2, num_pts):
            x = cx + lad * math.cos(theta)
            y = cy + lad * math.sin(theta)
            marker.points.append(Point(x=x, y=y, z=0.0))
        self.pub_lad.publish(marker)

    def vis_target(self, target):
        marker = Marker()
        marker.header.frame_id = 'base_link'
        marker.header.stamp = self.get_clock().now().to_msg()
        marker.ns = 'pp_target'
        marker.id = 2
        marker.type = Marker.SPHERE
        marker.action = Marker.ADD
        marker.scale.x = 0.25
        marker.scale.y = 0.25
        marker.scale.z = 0.25
        marker.color = ColorRGBA(r=1.0, g=0.2, b=1.0, a=1.0)
        marker.pose.orientation.w = 1.0
        marker.pose.position.x = float(target[0])
        marker.pose.position.y = float(target[1])
        marker.pose.position.z = 0.05
        self.pub_target.publish(marker)

    def vis_str(self, target):
        marker = Marker()
        marker.header.frame_id = 'base_link'
        marker.header.stamp = self.get_clock().now().to_msg()
        marker.ns = 'pp_target_line'
        marker.id = 3
        marker.type = Marker.LINE_STRIP
        marker.action = Marker.ADD
        marker.scale.x = 0.04
        marker.color = ColorRGBA(r=1.0, g=0.6, b=0.0, a=1.0)
        marker.pose.orientation.w = 1.0

        cx, cy = CAR_POSITION
        marker.points.append(Point(x=float(cx), y=float(cy), z=0.02))
        marker.points.append(Point(x=float(target[0]), y=float(target[1]), z=0.02))
        self.pub_line.publish(marker)

    def main_loop(self):
        if not self.path:
            return

        # ★★★ 변경점: 가변 LAD 계산 함수 호출 ★★★
        lad = self.calculate_variable_lad()
        target = self.find_target(lad)
        steering_angle = self.compute_steering(target)

        # publish steering angle in degrees
        steering_angle_deg = math.degrees(steering_angle)
        steering_angle_deg = float(np.clip(steering_angle_deg, -MAX_STEERING, MAX_STEERING))
        speed = float(self.SPEED)
        rpm = self.compute_speed2rpm(speed)
        self.drive(steering_angle_deg, rpm)

        self.vis_lad(lad)
        self.vis_target(target)
        self.vis_str(target)

def main(args=None):
    rclpy.init(args=args)
    node = PurePursuitControl()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()